config/src/main/java/com/yahoo/config/subscription/impl/ConfigSubscription.java

```diff
@@ -129,50 +159,52 @@ public boolean equals(Object o) {
         return false;
     }
 
-    void setConfigChanged(boolean changed) {
-        this.configChanged = changed;
-    }
-
-    void setGenerationChanged(boolean genChanged) {
-        this.generationChanged = genChanged;
-    }
 
     /**
      * Called from {@link ConfigSubscriber} when the changed status of this config is propagated to the clients
      */
-    public void resetChangedFlags() {
-        setConfigChanged(false);
-        setGenerationChanged(false);
+    public boolean isConfigChangedAndReset() {
+        ConfigState<T> prev = config.get();
+        while ( !config.compareAndSet(prev, prev.createUnchanged())) {
+            prev = config.get();
+        }
+        return prev.isConfigChanged();
     }
 
-    public boolean isConfigChanged() {
-        return configChanged;
+    void setConfig(Long generation, T config) {
+        this.config.set(new ConfigState<>(true, generation, true, config));
     }
 
-    public boolean isGenerationChanged() {
-        return generationChanged;
+    void setConfigIncGen(T config) {
+        ConfigState<T> prev = this.config.get();
+        while ( !this.config.compareAndSet(prev, new ConfigState<>(true, prev.getGeneration()+1, true, config))) {
+            prev = this.config.get();
+        }
     }
 
-    void setConfig(T config) {
-        this.config = config;
+    void setConfigIfChangedIncGen(T config) {
+        ConfigState<T> prev = this.config.get();
+        while (!this.config.compareAndSet(prev, new ConfigState<>(true, prev.getGeneration() + 1, !config.equals(prev.getConfig()), config))) {
```

Comment:
this loop also looks a bit iffy (hopefully unneeded)

Reply:
Agree, single writer anyway - Fixed